### PR TITLE
Setup automated testing for CUDA

### PR DESCRIPTION
**Pull request by dalg24**
@crtrott I know I promised Volta but I don't have the cycles yet.  I am still waiting on a couple nodes.
Anyway it looks like we get a segfault in the performance test with Clang and CUDA 10.1 does not compile in optimized with NVCC so I figured we could start here.

### CODE REVIEW

File: .jenkins [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                            sh 'ccache --show-stats'
+                        }
+                    }
+                }
+                stage("CUDA-10.1-NVCC") {
+                    agent {
+                        dockerfile {
+                            filename 'Dockerfile.nvcc'
+                            dir 'scripts/docker'
+                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:9.2-devel'
```
Comment (crtrott) [inline .jenkins]: why does this say 9.2-devel?

File: .jenkins [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                            sh 'ccache --show-stats'
+                        }
+                    }
+                }
+                stage("CUDA-10.1-NVCC-DEBUG") {
+                    agent {
+                        dockerfile {
+                            filename 'Dockerfile.nvcc'
+                            dir 'scripts/docker'
+                            additionalBuildArgs '--build-arg BASE=nvidia/cuda:9.2-devel'
```
Comment (crtrott) [inline .jenkins]: why does this say 9.2?

File: .jenkins (hunk 1 above)
Comment (crtrott) [reply .jenkins]: Is it just the base container?

File: .jenkins (hunk 2 above)
Comment (dalg24) [reply .jenkins]: Fixed

File: .jenkins (hunk 1 above)
Comment (dalg24) [reply .jenkins]: Yes

**Comment by dalg24** [comment]
@crtrott builds are now running on Volta.
Would you please look into CUDA 10.1 not building in release mode?

**Comment by masterleinad** [comment]
Looks like
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaGetDeviceCount( & m_cudaDevCount ) error( cudaErrorInsufficientDriver): CUDA driver version is insufficient for CUDA runtime version /var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:204
```
is the main problem here.

**Comment by nmhamster** [comment]
What machine is this happening on?

**Comment by dalg24** [comment]
> Looks like
> 
> ```
> terminate called after throwing an instance of 'std::runtime_error'
>   what():  cudaGetDeviceCount( & m_cudaDevCount ) error( cudaErrorInsufficientDriver): CUDA driver version is insufficient for CUDA runtime version /var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:204
> ```
> 
> is the main problem here.

I am not sure where you looked.  The build issue that needs attention is [here](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/10/pipeline)

**Comment by masterleinad** [comment]
I was looking [here](https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/9/console). Maybe just another problem then.

**Comment by dalg24** [comment]
> I was looking [here](https://cloud.cees.ornl.gov/jenkins-ci/job/Kokkos/9/console). Maybe just another problem then.

That's an old build.  There was an issue with one of the two nodes I bootstrapped and has been fixed.

**Comment by masterleinad** [comment]
Ah, sorry for the noise then.

**Comment by crtrott** [comment]
The error on the release build is a bad driver version. I.e. on that particular machine it looks like CUDA 10.1 got installed without the necessary upgrade of the driver:

```
./KokkosCore_UnitTest_Cuda
terminate called after throwing an instance of 'std::runtime_error'
  what():  cudaGetDeviceCount( & m_cudaDevCount ) error( cudaErrorInsufficientDriver): CUDA driver version is insufficient for CUDA runtime version
```

**Comment by dalg24** [comment]
@crtrott You looked at an older build.  The driver issue has been fixed.  You want to look at [this one](https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/10/pipeline)
```
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_TaskQueueMultiple_impl.hpp: In instantiation of 'void Kokkos::Impl::TaskQueueMultiple<ExecSpace, MemorySpace>::Destroy::destroy_shared_allocation() [with ExecSpace = Kokkos::Cuda; MemorySpace = Kokkos::CudaUVMSpace]':
/var/jenkins/workspace/Kokkos/core/src/Cuda/Kokkos_Cuda_Task.cpp:58:16:   required from here
/var/jenkins/workspace/Kokkos/core/src/impl/Kokkos_TaskQueueMultiple_impl.hpp:59:16: error: cannot call member function 'Kokkos::Impl::TaskQueueMultiple<ExecSpace, MemorySpace>::team_queue_type& Kokkos::Impl::TaskQueueMultiple<ExecSpace, MemorySpace>::get_team_queue(int) [with ExecSpace = Kokkos::Cuda; MemorySpace = Kokkos::CudaUVMSpace; Kokkos::Impl::TaskQueueMultiple<ExecSpace, MemorySpace>::team_queue_type = Kokkos::Impl::TaskQueueMultiple<Kokkos::Cuda, Kokkos::CudaUVMSpace>]' without object
   m_queue->get_team_queue(0).~TaskQueueMultiple();
   ~~~~~~~~~~~~~^~~
/var/jenkins/workspace/Kokkos/Makefile.targets:48: recipe for target 'Kokkos_Cuda_Task.o' failed
make[1]: *** [Kokkos_Cuda_Task.o] Error 1
make[1]: Leaving directory '/var/jenkins/workspace/Kokkos/build/core'
Makefile:12: recipe for target 'kokkoslib-test' failed
make: *** [kokkoslib-test] Error 2
```

**Comment by crtrott** [comment]
Whats the GCC version on this? I can't reproduce it right now on my box with CUDA 10.1. And the error clearly indicates a compiler issue since m_queue->get_team_queue(0) is calling a function on a concrete object.

**Comment by crtrott** [comment]
Lets get this in for now, and fix the RDC issue separately.

**Pull request merged**